Remove Tensor.h, TensorMethods.h from src/core.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27086 Remove Tensor.h, TensorMethods.h from src/core.**

Checked in generated files are a major source of merge conflicts. This is a commandeer of #25031.

Unfortunately, this patch is pretty involved. The general concept is to excise Caffe's dependency on ATen/core entirely. In practice, this means eliminating includes of `ATen/core/Tensor.h` and also all occurrences of IValues (who mention Tensor in their public interface.) This comes in several steps:

1. Modify all existing `CAFFE2_IS_XPLAT_BUILD` ifdefs to also consider `C10_MOBILE`. So basically, bring the OSS Caffe2 build and the XPLAT build in line. That eliminates a bunch of dependencies on TensorMethods.h. However, it's not enough, we need to...
2. Delete "c10 experimental" from the mobile build (cc @smessmer ) as empirically they are built as part of caffe2 mobile. We also need to
3. Remove a few more occurrences of IValue that don't seem to be handled by xplat today, e.g., caffe2/operators/reduction_ops.h (this is probably because we don't actually ship these ops in xplat today)
4. Cleanup on GradMode (ifdef'ed out!) and Context registration (move that header/cpp file back to Caffe2)

Differential Revision: [D17687345](https://our.internmc.facebook.com/intern/diff/D17687345)

ezyang committed:
Remove Tensor.h, TensorMethods.h from src/core.

This is a major source of merge conflicts, and AFAICT isn't necessary anymore (it may have been necessary for some mobile build stuff in the past).

[ghstack-poisoned]

ezyang committed:
Update on "Remove Tensor.h, TensorMethods.h from src/core."


This is a major source of merge conflicts, and AFAICT isn't necessary anymore (it may have been necessary for some mobile build stuff in the past).

[ghstack-poisoned]

ezyang committed:
Add missing reference to OpsAlreadyMovedToC10 on "Remove Tensor.h, TensorMethods.h from src/core."


This is a major source of merge conflicts, and AFAICT isn't necessary anymore (it may have been necessary for some mobile build stuff in the past).

[ghstack-poisoned]

ezyang committed:
Update on "Remove Tensor.h, TensorMethods.h from src/core."


This is a major source of merge conflicts, and AFAICT isn't necessary anymore (it may have been necessary for some mobile build stuff in the past).

This is a commandeer of #25031

Differential Revision: [D17687345](https://our.internmc.facebook.com/intern/diff/D17687345)

[ghstack-poisoned]

ezyang commented:
I believe the plan on record for Caffe2 Android build is that it should not need any deps on ATen/core

ezyang committed:
Update on "Remove Tensor.h, TensorMethods.h from src/core."


This is a major source of merge conflicts, and AFAICT isn't necessary anymore (it may have been necessary for some mobile build stuff in the past).

This is a commandeer of #25031

Differential Revision: [D17687345](https://our.internmc.facebook.com/intern/diff/D17687345)

[ghstack-poisoned]

ezyang committed:
Update on "Remove Tensor.h, TensorMethods.h from src/core."


This is a major source of merge conflicts, and AFAICT isn't necessary anymore (it may have been necessary for some mobile build stuff in the past).

This is a commandeer of #25031

Differential Revision: [D17687345](https://our.internmc.facebook.com/intern/diff/D17687345)

[ghstack-poisoned]

ezyang committed:
Update on "Remove Tensor.h, TensorMethods.h from src/core."


This is a major source of merge conflicts, and AFAICT isn't necessary anymore (it may have been necessary for some mobile build stuff in the past).

This is a commandeer of #25031

Differential Revision: [D17687345](https://our.internmc.facebook.com/intern/diff/D17687345)

[ghstack-poisoned]

ezyang commented on caffe2/core/context_base.cc:
```diff
@@ -1,5 +1,20 @@
-#include "context_base.h"
+#include <caffe2/core/context_base.h>
```
Code movement only (except for header locations update)

ezyang commented on caffe2/core/context_base.h:
```diff
@@ -1,7 +1,168 @@
+
 #include "caffe2/core/common.h"
 #include "caffe2/core/logging.h"
 #include "caffe2/proto/caffe2_pb.h"
```
Code movement only, except the headers that were previously in `caffe/core/context_base.h` are now inlined directly in the file.

ezyang commented on caffe2/core/export_c10_op_to_caffe2.h:
```diff
@@ -1,13 +1,16 @@
 
+#include <c10/macros/Macros.h>
+#include <c10/util/Registry.h>
+#include "caffe2/core/operator.h"
```
Some downstream deps (not in xplat, apparently?) seem to rely on this being in scope, so I whisked it out of the ifdef.

zou3519 commented on caffe2/CMakeLists.txt:
```diff
@@ -5,6 +5,7 @@ include(../cmake/Codegen.cmake)
 
 # ---[ ATen build
 if (INTERN_BUILD_ATEN_OPS)
+  message(WARNING "NOT only building core")
```
What does it mean to "NOT only [build] core"?

ezyang replied:
This is old from greg's patch, it can be deleted

zou3519 commented on caffe2/core/export_c10_op_to_caffe2.h:
```diff
@@ -1,13 +1,16 @@
+
 // TODO Also register c10 operators on mobile
-#if !defined(CAFFE2_IS_XPLAT_BUILD)
+#if !defined(CAFFE2_IS_XPLAT_BUILD) && !defined(C10_MOBILE)
```
Are there instances where we only consider CAFFE2_IS_XPLAT_BUIILD or C10_MOBILE? Could we delete one of these two in the future?

I remember @ljk53 was talking about consolidating the mobile build flags.

ezyang replied:
It looks like the answer is that, most of the time, we consider both as a pair. The only place this is not true is:

```
aten/src/ATen/core/op_registration/op_registration.cpp
2:#if !defined(CAFFE2_IS_XPLAT_BUILD)
                                                              
aten/src/ATen/core/op_registration/op_registration.h          
10:#if !defined(CAFFE2_IS_XPLAT_BUILD)                                      
99:        #if defined(CAFFE2_IS_XPLAT_BUILD)

aten/src/ATen/core/EnableNamedTensor.h
9:#if !defined(CAFFE2_IS_XPLAT_BUILD) && (!defined(C10_MOBILE) || defined(FEATURE_TORCH_MOBILE))
```

So agreed here.

ljk53 replied:
I think semantically CAFFE2_IS_XPLAT_BUILD should be equivalent to "C10_MOBILE && ! FEATURE_TORCH_MOBILE" - it's xplat mobile build (thus C10_MOBILE) and caffe2 only as in current production (thus !FEATURE_TORCH_MOBILE). 

One argument @smessmer had to keep CAFFE2_IS_XPLAT_BUILD was that xplat has broader meaning than mobile, but I think the substantial difference is whether to build full torch (FEATURE_TORCH_MOBILE).

ezyang replied:
I don't really have a stone in the game, except that I don't want to make too many more changes to this diff as it stands right now :)

zou3519 reviewed (commented):
(not full review) why do we have to excise caffe2's dependency on aten/core to delete TensorMethods.h and TensorBody.h? Is this to clean up the code or does caffe2 not do codegen?

ezyang commented:
> why do we have to excise caffe2's dependency on aten/core to delete TensorMethods.h and TensorBody.h? Is this to clean up the code or does caffe2 not do codegen?

Good question. There are multiple viable approaches. This particular approach has the benefit that I don't  have to setup codegen for OSS Android build, or in Caffe2 xplat. The alternative is to not excise, and instead teach Caffe2 Android and xplat build how to do codegen. This seemed hard (to me, anyway), so I decided not to do it.

ezyang commented on aten/src/ATen/gen.py:
```diff
@@ -45,12 +45,7 @@
-else:
-    core_install_dir = os.path.join(options.source_path, 'core')
-
+core_install_dir = options.install_dir + '/core' if options.install_dir is not None else None
```
we should use os.path.join here

ezyang replied:
Actually, apparently, NOT using os.path.join is mandatory on Windows. With this change, you get:

```

15:15:11 CMake Error at caffe2/CMakeLists.txt:600 (add_library):
15:15:11   Syntax error in cmake code when parsing string
15:15:11 
15:15:11     C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/build/aten/src/ATen\core/TensorMethods.h
15:15:11 
15:15:11   Invalid character escape '\c'.
```

zou3519 replied:
That's... counterintuitive, but go for it. How did this work before?

gchanan replied:
ya I don't see how this worked before, then.

ezyang committed:
Update on "Remove Tensor.h, TensorMethods.h from src/core."


Checked in generated files are a major source of merge conflicts. This is a commandeer of #25031.

Unfortunately, this patch is pretty involved. The general concept is to excise Caffe's dependency on ATen/core entirely. In practice, this means eliminating includes of `ATen/core/Tensor.h` and also all occurrences of IValues (who mention Tensor in their public interface.) This comes in several steps:

1. Modify all existing `CAFFE2_IS_XPLAT_BUILD` ifdefs to also consider `C10_MOBILE`. So basically, bring the OSS Caffe2 build and the XPLAT build in line. That eliminates a bunch of dependencies on TensorMethods.h. However, it's not enough, we need to...
2. Delete "c10 experimental" from the mobile build (cc @smessmer ) as empirically they are built as part of caffe2 mobile. We also need to
3. Remove a few more occurrences of IValue that don't seem to be handled by xplat today, e.g., caffe2/operators/reduction_ops.h (this is probably because we don't actually ship these ops in xplat today)
4. Cleanup on GradMode (ifdef'ed out!) and Context registration (move that header/cpp file back to Caffe2)

Differential Revision: [D17687345](https://our.internmc.facebook.com/intern/diff/D17687345)

[ghstack-poisoned]

zou3519 reviewed (approved):
lgtm, but I don't have context on the caffe2 bits. The guards should be as fine as all the builds build.

ezyang committed:
Fix Windows again on "Remove Tensor.h, TensorMethods.h from src/core."


Checked in generated files are a major source of merge conflicts. This is a commandeer of #25031.

Unfortunately, this patch is pretty involved. The general concept is to excise Caffe's dependency on ATen/core entirely. In practice, this means eliminating includes of `ATen/core/Tensor.h` and also all occurrences of IValues (who mention Tensor in their public interface.) This comes in several steps:

1. Modify all existing `CAFFE2_IS_XPLAT_BUILD` ifdefs to also consider `C10_MOBILE`. So basically, bring the OSS Caffe2 build and the XPLAT build in line. That eliminates a bunch of dependencies on TensorMethods.h. However, it's not enough, we need to...
2. Delete "c10 experimental" from the mobile build (cc @smessmer ) as empirically they are built as part of caffe2 mobile. We also need to
3. Remove a few more occurrences of IValue that don't seem to be handled by xplat today, e.g., caffe2/operators/reduction_ops.h (this is probably because we don't actually ship these ops in xplat today)
4. Cleanup on GradMode (ifdef'ed out!) and Context registration (move that header/cpp file back to Caffe2)

Differential Revision: [D17687345](https://our.internmc.facebook.com/intern/diff/D17687345)

[ghstack-poisoned]

gchanan commented:
this seems kind of annoying to merge.  Perhaps you should break it up into:
1) separate caffe2 bits
2) codegen changes

You could also consider temporarily keeping the checked in files but ignoring them (so we don't break in-flight patches), but that might not be worth it.

ljk53 commented:
@ezyang I updated internal diff D16983025 to change fbcode to use the generated files. That diff depends on some change on this PR (core_tmp -> core) and this PR needs that diff to not break fbcode. We need land them together or you can merge my change into D17687345 or maybe you should consider keeping generated files for a while as suggested by @gchanan then after D16983025 is landed we can remove generated files.

ezyang commented:
> this seems kind of annoying to merge

Does that mean if I manage to merge it, all is well? :) Unlike other patches, this won't silently merge conflict with anyone else's stuff. I might just do it over the weekend.

@ljk53 I'll just squash the diffs together.

ezyang committed:
Update on "Remove Tensor.h, TensorMethods.h from src/core."


Checked in generated files are a major source of merge conflicts. This is a commandeer of #25031.

Unfortunately, this patch is pretty involved. The general concept is to excise Caffe's dependency on ATen/core entirely. In practice, this means eliminating includes of `ATen/core/Tensor.h` and also all occurrences of IValues (who mention Tensor in their public interface.) This comes in several steps:

1. Modify all existing `CAFFE2_IS_XPLAT_BUILD` ifdefs to also consider `C10_MOBILE`. So basically, bring the OSS Caffe2 build and the XPLAT build in line. That eliminates a bunch of dependencies on TensorMethods.h. However, it's not enough, we need to...
2. Delete "c10 experimental" from the mobile build (cc @smessmer ) as empirically they are built as part of caffe2 mobile. We also need to
3. Remove a few more occurrences of IValue that don't seem to be handled by xplat today, e.g., caffe2/operators/reduction_ops.h (this is probably because we don't actually ship these ops in xplat today)
4. Cleanup on GradMode (ifdef'ed out!) and Context registration (move that header/cpp file back to Caffe2)

Differential Revision: [D17687345](https://our.internmc.facebook.com/intern/diff/D17687345)

[ghstack-poisoned]

ezyang committed:
Update on "Remove Tensor.h, TensorMethods.h from src/core."


Checked in generated files are a major source of merge conflicts. This is a commandeer of #25031.

Unfortunately, this patch is pretty involved. The general concept is to excise Caffe's dependency on ATen/core entirely. In practice, this means eliminating includes of `ATen/core/Tensor.h` and also all occurrences of IValues (who mention Tensor in their public interface.) This comes in several steps:

1. Modify all existing `CAFFE2_IS_XPLAT_BUILD` ifdefs to also consider `C10_MOBILE`. So basically, bring the OSS Caffe2 build and the XPLAT build in line. That eliminates a bunch of dependencies on TensorMethods.h. However, it's not enough, we need to...
2. Delete "c10 experimental" from the mobile build (cc @smessmer ) as empirically they are built as part of caffe2 mobile. We also need to
3. Remove a few more occurrences of IValue that don't seem to be handled by xplat today, e.g., caffe2/operators/reduction_ops.h (this is probably because we don't actually ship these ops in xplat today)
4. Cleanup on GradMode (ifdef'ed out!) and Context registration (move that header/cpp file back to Caffe2)

Differential Revision: [D17687345](https://our.internmc.facebook.com/intern/diff/D17687345)

[ghstack-poisoned]
